File: benchmark/rule110.cpp
#include <iostream>
#include <vector>

void print_cells(const std::vector<int>& cells) {
    for (int c : cells)
        std::cout << (c ? "x " : "  ");
    std::cout << "\n";
}

int rule_110(int left, int center, int right) {
    int pattern = (left << 2) | (center << 1) | right;
    return (pattern == 1 || pattern == 2 || pattern == 3 || pattern == 5 || pattern == 6) ? 1 : 0;
}

int main() {
    int size = 50;
    int generations = size;

    std::vector<int> cells(size, 0);
    cells[size - 1] = 1;

    print_cells(cells);

    for (int gen = 0; gen < generations; gen++) {
        std::vector<int> next_gen(size);
        for (int i = 0; i < size; i++) {
            int left = (i > 0) ? cells[i - 1] : 0;
            int center = cells[i];
            int right = (i < size - 1) ? cells[i + 1] : 0;
            next_gen[i] = rule_110(left, center, right);
        }
        print_cells(next_gen);
        cells = next_gen;
    }
    return 0;
}

File: benchmark/rule110.go
package main

import "fmt"

func printCells(cells []int) {
    for _, c := range cells {
        if c == 1 {
            fmt.Print("x ")
        } else {
            fmt.Print("  ")
        }
    }
    fmt.Println()
}

func rule110(left, center, right int) int {
    pattern := (left << 2) | (center << 1) | right
    switch pattern {
    case 1, 2, 3, 5, 6:
        return 1
    default:
        return 0
    }
}

func main() {
    size := 50
    generations := size

    cells := make([]int, size)
    cells[size-1] = 1

    printCells(cells)

    for gen := 0; gen < generations; gen++ {
        nextGen := make([]int, size)
        for i := 0; i < size; i++ {
            left, center, right := 0, cells[i], 0
            if i > 0 {
                left = cells[i-1]
            }
            if i < size-1 {
                right = cells[i+1]
            }
            nextGen[i] = rule110(left, center, right)
        }
        printCells(nextGen)
        cells = nextGen
    }
}

